### PR TITLE
[AIRFLOW-4797] Fix zombie detection

### PR DESCRIPTION
**Pull request by seelmann**
### Jira

- [X] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title.
  - https://issues.apache.org/jira/browse/AIRFLOW-4797

### Description

- [X] Here are some details about my PR, including screenshots of any UI changes:

Moved query to fetch zombies from `DagFileProcessorManager` to `DagBag` class. Changed query to only look for DAGs of the current DAG bag. The query now uses index `ti_dag_state` instead of `ti_state`. Removed no longer required `zombies` parameters from many function signatures.
 
The query is now executed on every call to `DagBag.kill_zombies` which is called when the DAG file is processed which frequency depends on `scheduler_heartbeat_sec` and `processor_poll_interval` (AFAIU). The query is faster than the previous one (see also stats below). It's also negligible IMHO because during DAG file processing many other queries (DAG runs and task instances are created, task instance dependencies are checked) are executed.

Tested on our staging environment (patch applied to Airflow 1.10.3), zombie detection works fine, database load is unchanged. Here some stats from `pg_stat_statements`, the branch run there for 4 hours: The new query (1st line) is faster but is likely called more frequently. The 2nd line shows stats of the old query.
```
select calls,mean_time,max_time,rows from pg_stat_statements where query like '%task_instance JOIN job%' and query like '%latest_heartbeat%';
  calls   |     mean_time      |  max_time   | rows 
----------+--------------------+-------------+------
    55416 | 0.0260821553522449 |    5.509762 |   29
 71969011 |  0.575755060854888 | 1078.895322 | 2377
```

Closed https://github.com/apache/airflow/pull/5420 in favour of this.

### Tests

- [X] My PR adds the following unit tests

### Commits

- [X] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [ ] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain docstrings that explain what it does
  - If you implement backwards incompatible changes, please leave a note in the [Updating.md](https://github.com/apache/airflow/blob/master/UPDATING.md) so we can assign it to a appropriate release

### Code Quality

- [X] Passes `flake8`

### CODE REVIEW

File: airflow/models/dagbag.py [hunk 1]
```diff
@@ -268,35 +270,46 @@ def process_file(self, filepath, only_if_updated=True, safe_mode=True):
...
+            configuration.getint('scheduler', 'scheduler_zombie_task_threshold'))
+        limit_dttm = timezone.utcnow() - timedelta(
+            seconds=zombie_threshold_secs)
+        self.log.debug("Failing jobs without heartbeat after %s", limit_dttm)
+
+        tis = (
+            session.query(TI)
+            .join(LJ, TI.job_id == LJ.id)
+            .filter(TI.state == State.RUNNING)
+            .filter(TI.dag_id.in_(self.dags))
```
Comment (ashb) [inline airflow/models/dagbag.py:296]: This works (I guess it must) even though dags is a dict?

**Review by ashb — APPROVED** [review]
This seems like a nice change - not having to pass that zombies list around def makes the code easier to follow.

I guess the new query is faster as we now include the dag ids in there so it's using the index better?

**Comment by ashb** [comment]
We're going to have an 1.10.4RC3 - how confident are you that it works? Should I include it in there?

**Comment by seelmann** [comment]
Yes, I assume the new query is faster because of the two components (dag and state) the number of candidates is smaller. However it only run for a few hours so the stats may change over time.

I'm very confident that it works. I'd deploy it to prod but I'm on vacation next week and don't want to leave my colleagues with another hotfix because the first attempt (https://github.com/apache/airflow/pull/5420) is already there, but afterwards I'll apply. So yes, please include in the next 1.10.4 RC, I can provide the adapted patch if it helps.

**Comment by ashb** [comment]
If you have a patch for 1.10 too yes please, that would save me some time!

**Comment by seelmann** [comment]
I'll send the patch tomorrow

**Comment by seelmann** [comment]
Patch for v1-10-stable is here: https://github.com/seelmann/incubator-airflow/commit/d48671c515e9dcf8e10e527557fb11f33350ff5e

But the cherry-pick is easy, only two conflicts with imports.

**Pull request merged**